File: core/src/main/java/heast/core/network/pipeline/PacketDecryptor.java
package heast.core.network.pipeline;

import heast.core.logging.IO;
import heast.core.security.AES;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToMessageDecoder;

import javax.crypto.SecretKey;
import java.util.List;

public final class PacketDecryptor extends MessageToMessageDecoder<ByteBuf> {

    private final SecretKey key;

    public PacketDecryptor(SecretKey key) {
        this.key = key;
    }

    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out) {
        var readable = msg.readableBytes();
        var bytes = new byte[readable];
        msg.readBytes(bytes);

        var buf = ctx.alloc().heapBuffer();
        var decrypted = AES.decrypt(key, bytes);
        buf.writeBytes(decrypted);
        out.add(buf);
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        IO.error.println("Error while decrypting packet: " + cause.getMessage());
        cause.printStackTrace();
    }
}

File: Core/Network/Pipeline/PacketDecryptor.cs
using System.Security;
using System.Security.Cryptography;
using DotNetty.Buffers;
using DotNetty.Codecs;
using DotNetty.Transport.Channels;

namespace Core.Network.Pipeline; 

public class PacketDecryptor : MessageToMessageDecoder<IByteBuffer> {
    
    private Aes KeyPair { get; }
    
    public PacketDecryptor(Aes keyPair) {
        KeyPair = keyPair;
    }

    protected override void Decode(IChannelHandlerContext context, IByteBuffer message, List<object> output) {
        var rb = message.ReadableBytes;
        var bytes = new byte[rb];
        message.ReadBytes(bytes);

        var buf = context.Allocator.HeapBuffer();
        var decrypted = new Span<byte>();
        var success = KeyPair.TryDecryptCfb(bytes, KeyPair.IV, decrypted, out _, PaddingMode.PKCS7);
        if (!success) throw new SecurityException("Failed to decrypt packet!");

        buf.WriteBytes(decrypted.ToArray());
        output.Add(buf);
    }

    public override void ExceptionCaught(IChannelHandlerContext context, Exception exception) {
        context.DisconnectAsync();
        Console.WriteLine($"Client kicked: {exception.Message}");
    }
}